experimental/collator/src/lib.rs

```diff
@@ -0,0 +1,4649 @@
+        // make it greater than zero when there is only one
+        // combining character and we don't do the trie lookup.
+        CharacterAndClass(u32::from(c) | (0xFF << 24))
+    }
+    pub fn character(&self) -> char {
+        unsafe { char::from_u32_unchecked(self.0 & 0xFFFFFF) }
+    }
+    pub fn ccc(&self) -> CanonicalCombiningClass {
+        CanonicalCombiningClass((self.0 >> 24) as u8)
+    }
+    // XXX need better naming here.
+    pub fn set_ccc(&mut self, ccc: &CodePointTrie<CanonicalCombiningClass>) {
+        debug_assert_eq!(self.0 >> 24, 0xFF, "This method has already been called!");
+        let scalar = self.0 & 0xFFFFFF;
+        self.0 = ((ccc.get(scalar).0 as u32) << 24) | scalar;
+    }
+}
+
+// This trivial function exists as a borrow check helper.
+#[inline(always)]
+fn sort_slice_by_ccc<'data>(
+    slice: &mut [char],
+    ccc: &CodePointTrie<'data, CanonicalCombiningClass>,
+) {
+    slice.sort_by_key(|cc| ccc.get(u32::from(*cc)));
+}
+
+fn data_ce_to_primary(data_ce: u64, c: char) -> u32 {
+    // Collation::getThreeBytePrimaryForOffsetData
+    let p = (data_ce >> 32) as u32; // three-byte primary pppppp00
+    let lower32 = data_ce as u32 as i32; // base code point b & step s: bbbbbbss (bit 7: isCompressible)
+    let mut offset = ((u32::from(c) as i32) - (lower32 >> 8)) * (lower32 & 0x7F); // delta * increment
+    let is_compressible = (lower32 & 0x80) != 0;
+    // Collation::incThreeBytePrimaryByOffset
+    offset += (((p >> 8) & 0xFF) as i32) - 2;
+    let mut primary = (((offset % 254) + 2) as u32) << 8;
+    offset /= 254;
+    // Same with the second byte,
+    // but reserve the PRIMARY_COMPRESSION_LOW_BYTE and high byte if necessary.
+    if is_compressible {
+        offset += (((p >> 16) & 0xFF) as i32) - 4;
+        primary |= (((offset % 251) + 4) as u32) << 16;
+        offset /= 251;
+    } else {
+        offset += (((p >> 16) & 0xFF) as i32) - 2;
+        primary |= (((offset % 254) + 2) as u32) << 16;
+        offset /= 254;
+    }
+    primary | ((p & 0xFF000000) + ((offset as u32) << 24))
+}
+
+/// Iterator that transforms an iterator over `char` into an iterator
+/// over `CollationElement` with a tailoring.
+/// Not a real Rust iterator: Instead of `None` uses `NO_CE` to indicate
+/// end of iteration to optimize comparison.
+struct CollationElements<'data, I>
+where
+    I: Iterator<Item = char>,
+{
+    iter: I,
+    pending: SmallVec<[CollationElement; 6]>, // TODO Figure out good length
+    // Two kinds of prefixes:
+    // Prefixes that contain a single starter
+    // Prefixes that contain a starter followed by either U+3099 or U+309A
+    // Last-pushed is at index 0 and previously-pushed at index 1
+    prefix: [char; 2],
+    // Invariant: the last char of the suffix is a potentially-precomposed
+    // starter unless iter has been exhausted.
+    suffix: SmallVec<[char; 10]>, // TODO Figure out good length; longest contraction suffix is 7 characters long
```

Comment:
suggestion: rename and/or document `suffix` to indicate its usage

IIUC, you use `suffix` as a queue to buffer the chars you are iterating over.  If it's serving that general purpose most of the time, would `buffer` or `queue` work just as well, or is there some tight and exclusive connection to `prefix` that I'm missing?

Along those lines, repeating the same question below in the normalizer code about whether a circular buffer could help (if you can feel confident on a specific finite capacity) fits here too.  Not sure how well that would help, but enqueuing by appending (`.push()`) and dequeuing  by shifting everything down by one (`.remove(0)`) seems like room for improvement at the level we care about.

Reply:
Renamed `suffix` to `upcoming`.

Documented `upcoming` and explained why it doesn't need a move avoidance optimization (typically, it stays at length 1). Asserted the documented invariant. Noticed that the invariant didn't actually hold! Fixed that bug.

Added a move avoidance optimization for `pending`.